### ISSUE TITLE
CachedImage App Freeze and Memory Problems

### ISSUE THREAD

**Issue by john-larson** [luberda-molinet/FFImageLoading#867]
I am using a Xamarin Forms ListView to display images from the internet. (ListView has HasUnevenRows=true and no caching strategy defined, if of any relevance.)

These images are 400x400 max. jpeg images. They range between 24KB-145KB. I am showing 3 images in each row. Xamarin image views are 100x100 and I am using Samsung Galaxy S3 for testing (pixel ratio:2).

The problem is, when using CachedImage (FFImageLoding.Forms 2.2.9), I get out of memory exceptions as I start to scroll past about the middle of the list and some of the image views start to show empty. When I inspected, the size of these images are sometimes 30 KB, sometimes 90 KB. So size of the image is irrelevant.

I tried using CacheType=Disk, thinking that each seen image is cached in memory, but the result was the same

I tried setting DownSampleHeight and DownSampleWidth properties to 100. Then the out of memory exceptions went away and scroll became smoother. But the problem was that, when I was testing by scrolling constantly and quickly up and down the list, app freezed and never recovered. I tried again and the same result. I could not understand why, do you have an idea?

Another thing is, images are already of small size. With a Samsung Galaxy S8 (pixel ratio:4), there won't be any need to downsample anyway, so aside from the freeze, I feel like downsampling should not be necessary anyway.

So, I tried using the standard Image view of Xamarin Forms instead of CachedImage. It worked fine without any out of memory exceptions. (But since there is no downsampling here, I guess scrolling was not as smooth as the CachedImage with downsampling.)  And when I tested by scrolling constantly and quickly up and down the list, app did not freeze at all.

Bottom line is,

1. How come the standard Xamarin Forms Image view is not causing out of memory exceptions but CachedImage is, when not using downsampling?
2. What can be the reason of app freezing when using CachedImage with downsampling and doing constant and quick scrolling?

**Comment by daniel-luberda** [comment]
FFImageLoding.Forms 2.2.9 is old and unsupported version, current is 2.3.1. There was a huge list of fixes / changes between those versions. Did you try to use it?

**Comment by john-larson** [comment]
I just updated and tested again. Here are my observations:

1. CachedImage is again causing out of memory exceptions without the downsampling, as opposed to the standard Xamarin Forms Image view.
2. The freezing I mentioned when scrolling the list up and down quickly seems to have been fixed, which is great. Do you remember this as an issue fixed and what was causing this?

**Comment by daniel-luberda** [comment]
> CachedImage is again causing out of memory exceptions without the downsampling, as opposed to the standard Xamarin Forms Image view.

FFImageLoading's memory cache takes some RAM to operate correctly vs Xamarin.Forms doesn't use any memory caching at all. Without memory cache your scrolling experience would be worse. See this comparison: https://www.youtube.com/watch?v=Dca3ksv8esU 

When you use a lot of different images, you should:
- enable largeheap in Android manifest for older devices: https://stackoverflow.com/a/23269892/5064986
- use downsampling to free some memory
- tune `MaxMemoryCacheSize` at app startup: `ImageService.Instance.Init(new Configuration { MaxMemoryCacheSize=[some value]})` (default is 20% of available app memory)
- Invalidate memory cache when large percent of images are invalidated

> The freezing I mentioned when scrolling the list up and down quickly seems to have been fixed, which is great. Do you remember this as an issue fixed and what was causing this?

Sorry, I don't remember.

**Comment by john-larson** [comment]
> FFImageLoading's memory cache takes some RAM to operate correctly vs Xamarin.Forms doesn't use any memory caching at all.

I tried using `CacheType=Disk` without downsampling. That should have made it equivalent to the Xamarin.Forms Image view right? But it did not make a difference at all in terms of out of memory exceptions. Why do you think this might be?

**Comment by daniel-luberda** [comment]
I'll definitely need some sample repo to give any reasons. There might be a lot of things as maybe on older devices try to disable fade in / out animations, etc

**Comment by john-larson** [comment]
I tried to prepare an isolated reproduction project for testing and sent it to your email. I normally get the images via a url but I included the exact images in the project for convenience. 

It would be great if you could test it on a Samsung Galaxy S3 or a device with similar hardware if you could?

My questions are basically:

1. What causes it to throw "out of memory" exceptions when using `CacheType=Disk` and no downsampling as opposed to a regular Xamarin.Forms Image?
2. One other thing I noticed is that, when I scroll down very fast from the start, the end row(s) usually show the wrong images when using Xamarin.Form.Image. But using FFImageLoading.CachedImage does not do that, works perfect. Do you know what causes this? How does your library fix this? Because it should be unacceptable in terms of Xamarin.Forms.

By the way, I would like to thank you for such a great library, and for such prompt responses to questions :thumbsup:

**Comment by john-larson** [comment]
One more thing: I realized that I forgot to add `.DownsampleUseDipUnits = true;`  to the images.  Since Samsung Galaxy S3 pixel ratio is 2, after adding this, each image bitmap should now be 4 times bigger. But now I tried adding it, and fortunately it did not still throw "out of memory" exception so far.

**Comment by daniel-luberda** [comment]
@john-larson Thank you, it's always nice to hear that's useful for others. 

I debugged your sample and I couldn't see any memory leaks. On really old devices sometimes there are OOMs while swiping lists really fast, but it's not related to memory leaks, rather temporary "peak" memory usage. I don't know what's the cause, but I suppose somehow FFImageLoading decodes bitmaps in a higher rate. There are some things you can do to overcome that: 

- Set `LoadingDelay` of your `CachedImage` to some higher value like `200ms`
- On older devices use FFImageLoading custom config and set `SchedulerMaxParallelTasks` to some lower value (default is: `Math.Max(2, (int)(Environment.ProcessorCount / 2d))` 

> One other thing I noticed is that, when I scroll down very fast from the start, the end row(s) usually show the wrong images when using Xamarin.Form.Image. But using FFImageLoading.CachedImage does not do that, works perfect. Do you know what causes this? How does your library fix this? Because it should be unacceptable in terms of Xamarin.Forms.

- FFImageLoading tracks all the image loading tasks and if two tasks are assigned for the same view, the older one is cancelled. 
- BTW: This pattern will give you a large performance boost: https://github.com/luberda-molinet/FFImageLoading/wiki/Xamarin.Forms-Advanced#usage-in-listview-with-listviewcachingstrategyrecycleelement-enabled

**Comment by PatrLind** [comment]
I see similar problems i my application

**Comment by daniel-luberda** [comment]
This should be fixed with: https://github.com/luberda-molinet/FFImageLoading/commit/0d2f206502338774643e8ff7eb796e9273514ae8  (2.3.3 stable). Please reopen if still experiencing it.
